cranelift/filetests/src/function_runner.rs

```diff
@@ -1,96 +1,381 @@
-        Self { function, isa }
+impl SingleFunctionCompiler {
+    /// Build a [SingleFunctionCompiler] from a [TargetIsa]. For functions to be runnable on the
+    /// host machine, this [TargetISA] must match the host machine's ISA (see [with_host_isa]).
+    pub fn new(isa: Box<dyn TargetIsa>) -> Self {
+        let trampolines = HashMap::new();
+        Self { isa, trampolines }
     }
 
-    /// Build a function runner using the host machine's ISA and the passed flags
-    pub fn with_host_isa(function: Function, flags: settings::Flags) -> Self {
+    /// Build a [SingleFunctionCompiler] using the host machine's ISA and the passed flags.
+    pub fn with_host_isa(flags: settings::Flags) -> Self {
         let builder = host_isa_builder().expect("Unable to build a TargetIsa for the current host");
         let isa = builder.finish(flags);
-        Self::new(function, isa)
+        Self::new(isa)
     }
 
-    /// Build a function runner using the host machine's ISA and the default flags for this ISA
-    pub fn with_default_host_isa(function: Function) -> Self {
+    /// Build a [SingleFunctionCompiler] using the host machine's ISA and the default flags for this
+    /// ISA.
+    pub fn with_default_host_isa() -> Self {
         let flags = settings::Flags::new(settings::builder());
-        Self::with_host_isa(function, flags)
-    }
-
-    /// Compile and execute a single function, expecting a boolean to be returned; a 'true' value is
-    /// interpreted as a successful test execution and mapped to Ok whereas a 'false' value is
-    /// interpreted as a failed test and mapped to Err.
-    pub fn run(&self) -> Result<(), String> {
-        let func = self.function.clone();
-        if !(func.signature.params.is_empty()
-            && func.signature.returns.len() == 1
-            && func.signature.returns.first().unwrap().value_type.is_bool())
-        {
-            return Err(String::from(
-                "Functions must have a signature like: () -> boolean",
-            ));
+        Self::with_host_isa(flags)
+    }
+
+    /// Compile the passed [Function] to a [CompiledFunction]. This function will:
+    ///  - check that the default ISA calling convention is used (to ensure it can be called)
+    ///  - compile the [Function]
+    ///  - compile a [Trampoline] for the [Function]'s signature (or used a cached [Trampoline];
+    ///    this makes it possible to call functions when the signature is not known until runtime.
+    pub fn compile(&mut self, function: Function) -> Result<CompiledFunction, CompilationError> {
+        if function.signature.call_conv != self.isa.default_call_conv() {
+            return Err(CompilationError::InvalidTargetIsa);
+        }
+
+        // Compile the function itself.
+        let code_page = compile(function.clone(), self.isa.as_ref())?; // TODO how can we avoid this function.clone()? Context requires ownership of the function...
```

Comment:
Could we clone the function's signature before, since it's the only field that's used after (and it's cheap), and then pass the ownership to `compile`?

Reply:
Good idea... done.